### PR TITLE
Refacotor optimizer step logic

### PR DESCRIPTION
**Pull request by vwxyzjn**
The current implementation is problematic when not doing gradient truncation. The following filediff shows the lines that are needed to add.

```diff
     if self.truncate_grads:
         if self.multi_gpu:
             self.optimizer.synchronize()
             self.scaler.unscale_(self.optimizer)
             nn.utils.clip_grad_norm_(self.model.parameters(), self.grad_norm)
             with self.optimizer.skip_synchronize():
                 self.scaler.step(self.optimizer)
                 self.scaler.update()
         else:
             self.scaler.unscale_(self.optimizer)
             nn.utils.clip_grad_norm_(self.model.parameters(), self.grad_norm)
             self.scaler.step(self.optimizer)
             self.scaler.update()
     else:
-        self.scaler.step(self.optimizer)
-        self.scaler.update()
+        if self.multi_gpu:
+            with self.optimizer.skip_synchronize():
+                self.scaler.step(self.optimizer)
+                self.scaler.update()
+        else:
+            self.scaler.step(self.optimizer)
+            self.scaler.update()
```

But this is pretty ugly... so I refactored it to 

```
        if self.multi_gpu:
            self.optimizer.synchronize()
        
        if self.truncate_grads:
            self.scaler.unscale_(self.optimizer)
            nn.utils.clip_grad_norm_(self.model.parameters(), self.grad_norm)

        if self.multi_gpu:
            with self.optimizer.skip_synchronize():
                self.scaler.step(self.optimizer)
                self.scaler.update()
        else:
            self.scaler.step(self.optimizer)
            self.scaler.update()
```

### CODE REVIEW

**Comment by Denys88** [comment]
Yeah I even had a comment about it in the code :)
Maybe lets remove mixed precision too? we canget rid of scaler. Anyway it doesnt help.
Do you want to apply this change before or after horovod replacement?

**Comment by vwxyzjn** [comment]
Maybe it's probably better to merge this PR as is? We can then remove mixed-precision in a PR that involves docs change and could impact downstream projects such as isaacgymenvs.

**Comment by Denys88** [comment]
Okay! Ill review all other prs after my work in the evening.

**Comment by markelsanz14** [comment]
@Denys88 do we know how much removing mixed precision will affect different model sizes? I agree in most cases it won't help, mainly when we have 2-5 layer NNs, but it might speed up training with larger NN sizes.

**Comment by vwxyzjn** [comment]
Hey @Denys88, @ViktorM @markelsanz14 and I had a quick chat about this, and maybe it's worth it to keep the mixed-precision option? I did a quick benchmark and found virtually no performance differences

image


However, as @markelsanz14 suggested maybe in large NNs mixed-precision could offer benefits.

**Comment by markelsanz14** [comment]
Yeah, I wouldn't put more work into it for now, but removing it altogether seems a bit drastic. I feel like it can be useful in the future with larger NN architectures.

**Comment by vwxyzjn** [comment]
More data on this, removing the scaler seems to make it ~5% faster (178 seconds vs 186 seconds):

image

**Comment by Denys88** [comment]
could you check it on some heave networks, including transformers.
I have my IG fork https://github.com/Denys88/IsaacGymEnvs with some fat neural networks like openai embedding from the paper or just transformer.
!https://github.com/Denys88/IsaacGymEnvs/blob/main/isaacgymenvs/cfg/train/AntPPO_tr.yaml
